cmd/registry/core/lint-openapi.go

```diff
@@ -0,0 +1,157 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+
+	"github.com/apigee/registry/rpc"
+	linter "github.com/googleapis/gnostic/metrics/lint"
+	"google.golang.org/protobuf/proto"
+	yaml "gopkg.in/yaml.v3"
+)
+
+// NewLintFromOpenAPI runs the API linter and returns the results.
+func NewLintFromOpenAPI(name string, b []byte) (*rpc.Lint, error) {
+	// create a tmp directory
+	root, err := ioutil.TempDir("", "registry-openapi-")
+	if err != nil {
+		return nil, err
+	}
+	// log.Printf("running in %s", root)
+	name = filepath.Base(name)
+	// whenever we finish, delete the tmp directory
+	defer os.RemoveAll(root)
+	// write the file to the temp directory
+	spec, err := GUnzippedBytes(b)
+	if err != nil {
+		return nil, err
+	}
+	err = ioutil.WriteFile(root+"/"+name, spec, 0644)
+	if err != nil {
+		return nil, err
+	}
+	// run the linter on the spec
+	lint := &rpc.Lint{}
+	lint.Name = name
+	// run the linter on the spec
+	lintFile, err := lintFileForOpenAPI(name, root)
+	if err != nil {
+		return nil, err
+	}
+	lint.Files = append(lint.Files, lintFile)
+	return lint, err
+}
+
+func lintFileForOpenAPI(path string, root string) (*rpc.LintFile, error) {
+	cmd := exec.Command("gnostic", path, "--linter-out=.")
+	cmd.Dir = root
+	_, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, err
+	}
+	b, err := ioutil.ReadFile(root + "/linter.pb")
+	if err != nil {
+		return nil, err
+	}
+	output := &linter.Linter{}
+	if err := proto.Unmarshal(b, output); err != nil {
+		return nil, err
+	}
+
+	nodeFinder, err := newNodeFinder(root + "/" + path)
+
+	problems := make([]*rpc.LintProblem, 0)
+	for _, message := range output.Messages {
+		problem := &rpc.LintProblem{}
+		problem.Message = message.Message
+		problem.Suggestion = message.Suggestion
+		keys := message.Keys
+		node, err := nodeFinder.findNode(keys)
+		if err == nil {
+			l := int32(node.Line)
+			c := int32(node.Column)
+			problem.Location = &rpc.LintLocation{}
+			problem.Location.StartPosition = &rpc.LintPosition{
+				LineNumber:   l,
+				ColumnNumber: c,
+			}
+			problem.Location.EndPosition = &rpc.LintPosition{
+				LineNumber:   l,
+				ColumnNumber: c + int32(len(node.Value)-1),
+			}
+		}
+		problems = append(problems, problem)
+	}
+	result := &rpc.LintFile{}
+	result.Problems = problems
+	return result, err
+}
+
+type nodeFinder struct {
+	node *yaml.Node
+}
+
+func newNodeFinder(filename string) (*nodeFinder, error) {
+	data, _ := ioutil.ReadFile(filename)
+	var node yaml.Node
+	err := yaml.Unmarshal(data, &node)
+	if err == nil {
+		return &nodeFinder{node: &node}, nil
+	}
+	return nil, err
```

Comment:
Nit: When guarding error cases it's more common for the success flow to continue after an error check.
```
if err := yaml.Unmarshal(data, &node); err != nil {
  return nil, err
}
return &nodeFinder{node: &node}, nil
```